x/cardchain/keeper/match.go

```diff
@@ -129,3 +143,86 @@ func (k Keeper) GetOutcome(ctx sdk.Context, match types.Match) (outcome types.Ou
 	}
 	return
 }
+
+func (k Keeper) TryHandleMatchOutcome(ctx sdk.Context, match *types.Match) error {
+	if match.PlayerA.Confirmed && match.PlayerB.Confirmed && match.ServerConfirmed {
+		return k.HandleMatchOutcome(ctx, match)
+	}
+	return nil
+}
+
+func (k Keeper) HandleMatchOutcome(ctx sdk.Context, match *types.Match) error {
+	// Evaluate Outcome
+	outcomes := []types.Outcome{match.Outcome, match.PlayerA.Outcome, match.PlayerB.Outcome}
+	slices.Sort(outcomes)
+	outcomes = slices.Compact(outcomes)
+	switch i := uint64(len(outcomes)); i {
+	case 1:
+		k.ReportServerMatch(ctx, match.Reporter, 1, true)
+	default:
+		k.ReportServerMatch(ctx, match.Reporter, i-1, false)
+	}
+
+	outcome, err := k.GetOutcome(ctx, *match)
+	match.Outcome = outcome
+
+	err = k.distributeCoins(ctx, match, outcome)
+	if err != nil {
+		return err
+	}
+
+	err = k.voteMatchCards(ctx, match)
+	if err != nil {
+		return err
+	}
+
+	// TODO: Votes
+	return nil
+}
+
+func (k Keeper) voteMatchCards(ctx sdk.Context, match *types.Match) error {
+	users, err := k.getMatchUsers(ctx, *match)
+	if err != nil {
+		return err
+	}
+	players := []*types.MatchPlayer{match.PlayerA, match.PlayerB}
+	for idx, player := range players {
+		// filter voted cards cards
+		otherPlayer := players[(idx+1)%2]
+		var otherPlayerCards []uint64
+		var cleanedVotes []*types.SingleVote
+		if match.ServerConfirmed {
+			otherPlayerCards = otherPlayer.PlayedCards
+		} else {
+			otherPlayerCards = otherPlayer.Deck
+		}
```

Comment:
do I understand this correctly:
If the playedCards are used or the whole deck for voting rights, depends on who is first to report? the client or the server?